File: cmd/ochakai/seed.go
package main

import (
	"archive/tar"
	"compress/gzip"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path"
	"sort"
	"strings"
	"time"

	"github.com/na0fu3y/ochakai/internal/domain"
)

// The empty base is the first thing a new deployment has, and the last
// thing anybody wants to look at. Catalogs solve it with connectors that
// crawl the warehouse; ochakai does not have those and will not
// (design doc 0081 §1: a harvester inside the server would need warehouse
// credentials it deliberately does not hold, and knowledge here is
// curated rather than collected).
//
// This is the third way. The warehouse is read by the operator's own
// client, with their own credentials — `bq query`, a Snowflake session,
// psql — and its answer is piped in here. ochakai turns those rows into
// draft concepts and prints them as an OKF bundle, which `ochakai import`
// writes like any other bundle:
//
//	bq query --max_rows=100000 --format=json --nouse_legacy_sql \
//	  'SELECT table_schema, table_name, column_name, data_type, is_nullable
//	     FROM `proj.dataset.INFORMATION_SCHEMA.COLUMNS` ORDER BY ordinal_position' |
//	  ochakai seed - | ochakai import -
//
// The row limit in that first line is not decoration. `bq query` prints
// the first 100 rows and says nothing about the rest, so a dataset with
// more columns than that seeds only the tables those rows reached — with
// no error anywhere, because nothing downstream can tell a truncated
// listing from a small one. seedTruncationNote below is what ochakai can
// say about it from where it stands.
//
// Nothing here connects to anything. No warehouse client is linked into
// the binary, no credential is read, and the same command serves any
// warehouse whose catalog can be spelled as the rows below — which is
// most of them, because INFORMATION_SCHEMA is the one thing they agree
// on.
//
// **What comes out is a draft, and that is the whole point.** A projected
// table entry is a skeleton somebody has to fill in: what the table is
// for, which column lies, when the load is late. Drafts land in the
// review queue where that gets written, and until a human rules on one it
// is marked as what it is (design doc 0069). A catalog's connector
// produces a thousand rows nobody vouches for; this produces a thousand
// questions, which is the honest shape of the same information.

// seedColumn is one row of the schema listing, spelled the way
// INFORMATION_SCHEMA.COLUMNS spells it. Unknown keys are ignored, so a
// `SELECT *` works and so does a hand-written export.
type seedColumn struct {
	Schema     string `json:"table_schema"`
	Table      string `json:"table_name"`
	Column     string `json:"column_name"`
	DataType   string `json:"data_type"`
	IsNullable string `json:"is_nullable"` // "YES" / "NO", as the standard has it
	Comment    string `json:"description"` // BigQuery's column description, when selected
}

// seedTable is the rows of one table, gathered.
type seedTable struct {
	schema, name string
	columns      []seedColumn
}

func cmdSeed(_ context.Context, args []string) error {
	fs := newBareFlagSet(
		"seed",
		"Usage: ochakai seed [flags] <file.json | ->\n\n"+
			"Turn a warehouse's own schema listing into a bundle of draft concepts,\n"+
			"printed as a tar.gz on stdout. One `BigQuery Table` concept per table,\n"+
			"with its columns as a markdown table and its address as `resource`.\n"+
			"Reads JSON rows (an array, or one object per line) with the\n"+
			"INFORMATION_SCHEMA.COLUMNS column names: table_schema, table_name,\n"+
			"column_name, data_type, is_nullable, and description where there is one.\n"+
			"Rows for the same table are gathered however they arrive.\n\n"+
			"ochakai connects to no warehouse and holds no credential of one: you run\n"+
			"the query, with your own client and your own identity, and pipe the answer\n"+
			"here. Every concept comes out as a draft, because a projected schema is a\n"+
			"skeleton somebody still has to say something about — which is what the\n"+
			"review queue is for.\n\n"+
			"Raise your client's row limit when you run that query: `bq query` prints\n"+
			"the first 100 rows unless --max_rows says otherwise, and a listing cut\n"+
			"there looks exactly like a small dataset from here.",
		"  bq query --max_rows=100000 --format=json --nouse_legacy_sql \\\n"+
			"    'SELECT table_schema, table_name, column_name, data_type, is_nullable\n"+
			"       FROM `proj.dataset.INFORMATION_SCHEMA.COLUMNS` ORDER BY ordinal_position' |\n"+
			"    ochakai seed - | ochakai import -\n"+
			"  ochakai seed columns.json > catalog.tar.gz   # look before you import\n"+
			"  ochakai seed --project my-proj --prefix warehouse/tables - | ochakai import -\n")
	project := fs.String("project", "", "the warehouse project or account the tables live in, written into each concept's `resource` address")
	prefix := fs.String("prefix", "tables", "the id prefix each concept is written under, e.g. tables/shop/orders")
	pos, err := exactArgs(fs, args, 1)
	if err != nil {
		return err
	}
	in := os.Stdin
	if pos[0] != "-" {
		f, err := os.Open(pos[0])
		if err != nil {
			return err
		}
		defer f.Close()
		in = f
	}
	cols, err := readSeedColumns(in)
	if err != nil {
		return err
	}
	if note := seedTruncationNote(len(cols)); note != "" {
		fmt.Fprintln(os.Stderr, "note:", note)
	}
	tables := gatherSeedTables(cols)
	if len(tables) == 0 {
		return fmt.Errorf("no rows with a table_name: is this the output of a SELECT over INFORMATION_SCHEMA.COLUMNS?")
	}
	if err := writeSeedBundle(os.Stdout, tables, *project, *prefix); err != nil {
		return err
	}
	fmt.Fprintf(os.Stderr, "seeded %d tables as drafts; pipe into `ochakai import -` to write them\n", len(tables))
	return nil
}

// clientDefaultRows is what `bq query` prints when nobody passes
// --max_rows, and the shape of the one failure this path has that nothing
// else can catch: the pipe succeeds, the bundle is well-formed, and the
// tables past the cut are simply not in it. A listing truncated there is
// indistinguishable from a small warehouse — except for its size, which is
// this exact number and rarely anything else by accident.
const clientDefaultRows = 100

// seedTruncationNote is that suspicion, said out loud once and never as an
// error: exactly 100 rows is a legitimate listing often enough that
// refusing it would be wrong, and a silent partial catalog is the thing
// somebody actually lost a day to.
func seedTruncationNote(rows int) string {
	if rows != clientDefaultRows {
		return ""
	}
	return fmt.Sprintf("the input is exactly %d rows, which is what a client's default row limit looks like:\n"+
		"      `bq query` prints %d unless you pass --max_rows, and drops the rest without saying so.\n"+
		"      If the schema is larger than that, rerun the query with the limit raised.", clientDefaultRows, clientDefaultRows)
}

// readSeedColumns accepts both shapes a warehouse client prints: one JSON
// array (bq --format=json), and one object per line (jq -c, psql's
// row_to_json). Telling them apart by the first non-space byte costs
// nothing and saves a flag nobody would remember.
func readSeedColumns(r io.Reader) ([]seedColumn, error) {
	data, err := io.ReadAll(r)
	if err != nil {
		return nil, err
	}
	trimmed := strings.TrimSpace(string(data))
	if trimmed == "" {
		return nil, fmt.Errorf("no input: pass the schema listing on stdin, or name a file")
	}
	if strings.HasPrefix(trimmed, "[") {
		var cols []seedColumn
		if err := json.Unmarshal([]byte(trimmed), &cols); err != nil {
			return nil, fmt.Errorf("reading the JSON array: %w", err)
		}
		return cols, nil
	}
	var cols []seedColumn
	dec := json.NewDecoder(strings.NewReader(trimmed))
	for {
		var c seedColumn
		if err := dec.Decode(&c); err != nil {
			if errors.Is(err, io.EOF) {
				break
			}
			return nil, fmt.Errorf("reading a JSON object per line: %w", err)
		}
		cols = append(cols, c)
	}
	return cols, nil
}

// gatherSeedTables groups columns by table, in the order the rows
// arrived: a listing ordered by ordinal_position keeps its column order,
// which is the order somebody reading the table expects.
func gatherSeedTables(cols []seedColumn) []seedTable {
	index := map[string]int{}
	var tables []seedTable
	for _, c := range cols {
		if c.Table == "" || c.Column == "" {
			continue
		}
		key := c.Schema + "." + c.Table
		i, ok := index[key]
		if !ok {
			i = len(tables)
			index[key] = i
			tables = append(tables, seedTable{schema: c.Schema, name: c.Table})
		}
		tables[i].columns = append(tables[i].columns, c)
	}
	sort.SliceStable(tables, func(i, j int) bool {
		if tables[i].schema != tables[j].schema {
			return tables[i].schema < tables[j].schema
		}
		return tables[i].name < tables[j].name
	})
	return tables
}

// writeSeedBundle prints the tables as an OKF bundle: one markdown
// document per table, at the path its id gives it (design doc 0046 §3.5).
func writeSeedBundle(w io.Writer, tables []seedTable, project, prefix string) error {
	gz := gzip.NewWriter(w)
	tw := tar.NewWriter(gz)
	for _, t := range tables {
		doc := seedDocument(t, project)
		name := path.Join(seedID(t, prefix) + ".md")
		if err := tw.WriteHeader(&tar.Header{
			Name: name, Mode: 0o644, Size: int64(len(doc)),
			ModTime: time.Time{}, Typeflag: tar.TypeReg,
		}); err != nil {
			return err
		}
		if _, err := tw.Write([]byte(doc)); err != nil {
			return err
		}
	}
	if err := tw.Close(); err != nil {
		return err
	}
	return gz.Close()
}

// seedID is the concept's address: the prefix, then the schema and table
// as path segments, so a dataset is a directory in the bundle and the
// browse view groups by it for free.
func seedID(t seedTable, prefix string) string {
	segments := []string{}
	for _, s := range []string{prefix, t.schema, t.name} {
		if s = domain.Normalize(strings.Trim(s, "/")); s != "" {
			segments = append(segments, s)
		}
	}
	return path.Join(segments...)
}

// seedDocument renders one table as an OKF document. It says only what
// the schema said — the prose is a prompt for the person who fills it in,
// not a description invented on their behalf. An empty description is
// deliberate: the frontmatter would otherwise carry a sentence nobody
// wrote and everybody would then trust.
func seedDocument(t seedTable, project string) string {
	b := &strings.Builder{}
	fmt.Fprintf(b, "---\ntype: %s\n", domain.TypeTables)
	if resource := seedResource(t, project); resource != "" {
		fmt.Fprintf(b, "resource: %s\n", resource)
	}
	fmt.Fprintf(b, "title: %s\nstatus: draft\n---\n\n", seedTitle(t))
	b.WriteString("Projected from the warehouse's own schema listing. **Nothing below was\n" +
		"written by a person yet** — what this table is for, which column lies, and\n" +
		"when the load is late are the reasons anybody will read this entry, and the\n" +
		"schema does not know any of them.\n\n")
	b.WriteString("| Column | Type | Null | Note |\n|---|---|---|---|\n")
	for _, c := range t.columns {
		null := ""
		if strings.EqualFold(c.IsNullable, "YES") {
			null = "yes"
		}
		fmt.Fprintf(b, "| `%s` | %s | %s | %s |\n",
			c.Column, c.DataType, null, strings.ReplaceAll(c.Comment, "|", "\\|"))
	}
	return b.String()
}

func seedTitle(t seedTable) string {
	if t.schema == "" {
		return t.name
	}
	return t.schema + "." + t.name
}

// seedResource is the table's address in the warehouse, in the spelling
// the demo bundle uses. Without --project the account is unknown and the
// key is left out rather than guessed: a resource that names the wrong
// project is worse than one that names none.
func seedResource(t seedTable, project string) string {
	if project == "" || t.name == "" {
		return ""
	}
	return "bigquery://" + project + "." + seedTitle(t)
}

File: internal/webui/static/js/seed.js
// The page's copy of `ochakai seed` (cmd/ochakai/seed.go): a dataset's
// INFORMATION_SCHEMA rows, read by the person as themselves, turned into
// one draft `BigQuery Table` concept per table (design doc 0148).
//
// It is a copy, and a copy drifts the day one side learns a rule the
// other did not. Both are held to one file for that reason —
// cmd/ochakai/testdata/seed-golden.json, which the Go test writes and
// internal/webui/jstest/seed.test.js reads — so the drift fails CI
// instead of reaching somebody's base. Change the Go side, rerun
// `go test ./cmd/ochakai -run TestSeedGolden -update`, and this file
// has to follow.

export const TABLE_TYPE = 'BigQuery Table';

// A source is `project.dataset`, or `project.dataset.table` for one
// table — `events_*` naming a date-sharded one. Spelled narrowly on
// purpose: every part is pasted into the query's FROM and WHERE, so
// what cannot be a BigQuery name is refused here rather than escaped.
const PROJECT = /^[a-z][a-z0-9.:-]*[a-z0-9]$/;
const DATASET = /^[A-Za-z0-9_]+$/;
const TABLE = /^[A-Za-z0-9_$-]+\*?$/;

export function parseSource(s) {
  const parts = s.trim().replace(/^`|`$/g, '').split('.');
  // A domain-scoped project (example.com:proj) carries a dot of its own.
  if (parts.length > 2 && parts[0].includes(':') === false && parts[1].includes(':')) {
    parts.splice(0, 2, parts[0] + '.' + parts[1]);
  }
  const [project, dataset, table = ''] = parts;
  if (parts.length < 2 || parts.length > 3 || !PROJECT.test(project || '') || !DATASET.test(dataset || '')
      || (table && !TABLE.test(table))) {
    throw new Error('取り込み元は project.dataset か project.dataset.table の形で書いてください(例: bigquery-public-data.thelook_ecommerce)');
  }
  return { project, dataset, table };
}

// schemaQuery is the one query the page runs. COLUMN_FIELD_PATHS is
// joined for the column descriptions BigQuery keeps there — the only
// prose in a schema a person wrote — and only at the top level, where
// field_path is the column itself.
export function schemaQuery({ project, dataset, table }) {
  const at = view => `\`${project}.${dataset}.INFORMATION_SCHEMA.${view}\``;
  let where = '';
  if (table.endsWith('*')) where = `\nWHERE STARTS_WITH(c.table_name, '${table.slice(0, -1)}')`;
  else if (table) where = `\nWHERE c.table_name = '${table}'`;
  return `SELECT c.table_schema, c.table_name, c.column_name, c.data_type, c.is_nullable, f.description
FROM ${at('COLUMNS')} c
LEFT JOIN ${at('COLUMN_FIELD_PATHS')} f
  ON f.table_name = c.table_name AND f.column_name = c.column_name AND f.field_path = c.column_name${where}
ORDER BY c.table_name, c.ordinal_position`;
}

// Byte order, as Go's string comparison has it: the golden file is
// sorted that way, and localeCompare would disagree with it.
const byName = (a, b) => (a.schema !== b.schema ? cmp(a.schema, b.schema) : cmp(a.name, b.name));
const cmp = (a, b) => (a < b ? -1 : a > b ? 1 : 0);
const str = v => (v === null || v === undefined ? '' : String(v));

// gather groups columns by table, in the order the rows arrived.
export function gather(rows) {
  const index = new Map();
  const tables = [];
  for (const r of rows) {
    const c = {
      schema: str(r.table_schema), table: str(r.table_name), column: str(r.column_name),
      dataType: str(r.data_type), isNullable: str(r.is_nullable), comment: str(r.description),
    };
    if (!c.table || !c.column) continue;
    const key = c.schema + '.' + c.table;
    if (!index.has(key)) {
      index.set(key, tables.length);
      tables.push({ schema: c.schema, name: c.table, columns: [], shards: [] });
    }
    tables[index.get(key)].columns.push(c);
  }
  return tables.sort(byName);
}

const SHARD = /^(.*[^0-9])([0-9]{8})$/;

function isDate(d) {
  const y = +d.slice(0, 4), m = +d.slice(4, 6), day = +d.slice(6, 8);
  const t = new Date(Date.UTC(y, m - 1, day));
  return t.getUTCFullYear() === y && t.getUTCMonth() === m - 1 && t.getUTCDate() === day;
}

// foldShards folds each set of two or more date-sharded tables into one
// entry named by the stem, with the latest shard's columns — unless the
// stem is itself a table's name. seed.go's foldShards says why.
export function foldShards(tables) {
  const named = new Set(tables.map(t => t.schema + '\x00' + t.name));
  const groups = new Map();
  tables.forEach((t, i) => {
    const m = SHARD.exec(t.name);
    if (!m || !isDate(m[2])) return;
    const k = t.schema + '\x00' + m[1];
    if (!groups.has(k)) groups.set(k, { schema: t.schema, stem: m[1], members: [] });
    groups.get(k).members.push(i);
  });
  const folded = new Set();
  const out = [];
  for (const [k, g] of groups) {
    if (g.members.length < 2 || named.has(k)) continue;
    let latest = g.members[0];
    const suffixes = [];
    for (const i of g.members) {
      folded.add(i);
      suffixes.push(tables[i].name.slice(g.stem.length));
      if (tables[i].name > tables[latest].name) latest = i;
    }
    out.push({ schema: g.schema, name: g.stem, columns: tables[latest].columns, shards: suffixes.sort(cmp) });
  }
  tables.forEach((t, i) => { if (!folded.has(i)) out.push(t); });
  return out.sort(byName);
}

// seedID is the concept's address: prefix, dataset, table.
export function seedID(t, prefix) {
  return [prefix, t.schema, t.name]
    .map(s => s.replace(/^\/+|\/+$/g, '').normalize('NFC'))
    .filter(Boolean).join('/');
}

function title(t) {
  const name = t.name + (t.shards.length ? '*' : '');
  return t.schema ? t.schema + '.' + name : name;
}

// seedDocument renders one table as an OKF document, byte for byte what
// seed.go's seedDocument writes.
export function seedDocument(t, project) {
  let b = `---\ntype: ${TABLE_TYPE}\n`;
  if (project && t.name) b += `resource: bigquery://${project}.${title(t)}\n`;
  b += `title: ${title(t)}\nstatus: draft\n---\n\n`;
  b += 'Projected from the warehouse\'s own schema listing. **Nothing below was\n'
    + 'written by a person yet** — what this table is for, which column lies, and\n'
    + 'when the load is late are the reasons anybody will read this entry, and the\n'
    + 'schema does not know any of them.\n\n';
  const n = t.shards.length;
  if (n) {
    const s = t.name;
    b += `**Date-sharded**: ${n} tables, \`${s}${t.shards[0]}\` to \`${s}${t.shards[n - 1]}\`, folded into this one\n`
      + `entry. Query them as \`${s}*\`, narrowing with \`_TABLE_SUFFIX\`; the columns\n`
      + 'below are the latest shard\'s.\n\n';
  }
  b += '| Column | Type | Null | Note |\n|---|---|---|---|\n';
  for (const c of t.columns) {
    const nul = c.isNullable.toUpperCase() === 'YES' ? 'yes' : '';
    b += `| \`${c.column}\` | ${c.dataType} | ${nul} | ${c.comment.replaceAll('|', '\\|')} |\n`;
  }
  return b;
}

// project is the whole projection: rows in, {id, title, document,
// columns, shards} out, in the order the bundle would list them.
export function project(rows, { project: proj = '', prefix = 'tables' } = {}) {
  return foldShards(gather(rows)).map(t => ({
    id: seedID(t, prefix), title: title(t), document: seedDocument(t, proj),
    columns: t.columns.length, shards: t.shards.length,
  }));
}
